File: src/main/java/com/github/morningzeng/toolset/utils/ScratchFileUtils.java
package com.github.morningzeng.toolset.utils;

import com.fasterxml.jackson.core.type.TypeReference;
import com.github.morningzeng.toolset.Constants;
import com.github.morningzeng.toolset.annotations.ScratchConfig;
import com.github.morningzeng.toolset.enums.OutputType;
import com.intellij.ide.scratch.ScratchFileService;
import com.intellij.ide.scratch.ScratchRootType;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Computable;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.newvfs.RefreshQueue;
import com.intellij.openapi.vfs.newvfs.RefreshSession;
import lombok.SneakyThrows;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.stream.Stream;

import static com.github.morningzeng.toolset.utils.JacksonUtils.IGNORE_TRANSIENT_AND_NULL;

/**
 * @author Morning Zeng
 * @since 2024-06-27
 */
public final class ScratchFileUtils {

    static final String ROOT_DIRECTORY = "Toolset";

    @SneakyThrows
    public static <T> void write(final T t, final TypeReference<T> typeReference) {
        if (Objects.isNull(t)) {
            return;
        }
        final ScratchConfig scratchConfig = getScratchConfig(typeReference);
        final String filename = String.join(Constants.DOT, scratchConfig.value(), scratchConfig.outputType().suffix());
        final VirtualFile file = findOrCreate(scratchConfig.directory(), filename);
        write(file, scratchConfig.outputType().serialize(t));
    }

    @SneakyThrows
    public static <T> void write(final String directory, String filename, final OutputType outputType, final T t) {
        filename = String.join(Constants.DOT, filename, outputType.suffix());
        final VirtualFile file = findOrCreate(directory, filename);
        write(file, outputType.serialize(t));
    }

    @SneakyThrows
    public static <T> void write(final VirtualFile file, final OutputType outputType, final T t) {
        write(file, outputType.serialize(t));
    }

    public static void write(final String name, final String content) {
        write(name, content, virtualFile -> {
        });
    }

    public static <T> void writeAndOpen(final Project project, @NotNull String name, final T t) {
        write(name, IGNORE_TRANSIENT_AND_NULL.toPrettyJson(t), scratchFile -> open(project, scratchFile));
    }

    public static void writeAndOpen(final Project project, @NotNull String name, final String content) {
        write(name, content, scratchFile -> open(project, scratchFile));
    }

    @SneakyThrows
    public static void write(final String filename, final String content, final Consumer<VirtualFile> consumer) {
        final VirtualFile scratchFile = findOrCreate(filename);
        write(scratchFile, content, consumer);
    }

    public static void write(final VirtualFile file, final String content) {
        write(file, content, virtualFile -> {
        });
    }

    public static void write(final VirtualFile file, final String content, final Consumer<VirtualFile> consumer) {
        ApplicationManager.getApplication().invokeLater(() -> {
            try {
                ApplicationManager.getApplication().runWriteAction(() -> {
                    try {
                        VfsUtil.saveText(file, content);
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }

                    RefreshSession rs = RefreshQueue.getInstance().createSession(true, true, null);
                    rs.addFile(file);
                    rs.launch();
                });
                consumer.accept(file);
            } catch (Exception ex) {
                throw new RuntimeException(ex);
            }
        });
    }

    public static String read(final VirtualFile file) {
        return ApplicationManager.getApplication().runReadAction((Computable<String>) () -> {
            try {
                return VfsUtil.loadText(file);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }

    @SneakyThrows
    public static String read(final String directory, final String filename) {
        return read(findOrCreate(directory, filename));
    }

    public static <T> T read(final Class<T> type) {
        final ScratchConfig scratchConfig = type.getAnnotation(ScratchConfig.class);
        if (Objects.isNull(scratchConfig)) {
            throw new IllegalArgumentException("ScratchConfig annotation is missing");
        }
        return read(scratchConfig.directory(), scratchConfig.value(), scratchConfig.outputType(), type);
    }

    public static <T> T read(final TypeReference<T> type) {
        final ScratchConfig scratchConfig = getScratchConfig(type);
        return read(scratchConfig.directory(), scratchConfig.value(), scratchConfig.outputType(), type);
    }

    public static <T> T read(final Type type) {
        final ScratchConfig scratchConfig = getScratchConfig(type);
        return read(scratchConfig.directory(), scratchConfig.value(), scratchConfig.outputType(), type);
    }

    public static <T> T read(final String directory, final String filename, final OutputType outputType, final Class<T> type) {
        final String read = read(directory, outputType.fullName(filename));
        if (StringUtil.isEmpty(read)) {
            return null;
        }
        return outputType.deserialize(read, type);
    }

    public static <T> T read(final String directory, final String filename, final OutputType outputType, final TypeReference<T> type) {
        final String read = read(directory, outputType.fullName(filename));
        if (StringUtil.isEmpty(read)) {
            return null;
        }
        return outputType.deserialize(read, type);
    }

    public static <T> T read(final String directory, final String filename, final OutputType outputType, final Type type) {
        final String read = read(directory, outputType.fullName(filename));
        if (StringUtil.isEmpty(read)) {
            return null;
        }
        return outputType.deserialize(read, type);
    }

    public static <T> T read(final VirtualFile file, final OutputType outputType, final Type type) {
        final String json = read(file);
        return outputType.deserialize(json, type);
    }

    public static <T> T read(final VirtualFile file, final OutputType outputType, final Class<T> type) {
        final String json = read(file);
        return outputType.deserialize(json, type);
    }

    public static <T> T read(final VirtualFile file, final OutputType outputType, final TypeReference<T> type) {
        final String json = read(file);
        return outputType.deserialize(json, type);
    }

    @SneakyThrows
    public static void openFile(final Project project, final String directory, final String filename) {
        final VirtualFile virtualFile = findOrCreate(directory, filename);
        open(project, virtualFile);
    }

    public static void childrenFile(final String dir, final Consumer<Stream<VirtualFile>> consumer) {
        ApplicationManager.getApplication().invokeLater(
                () -> {
                    final VirtualFile directory = directory(dir);
                    consumer.accept(Arrays.stream(directory.getChildren()));
                }
        );
    }

    public static VirtualFile findOrCreate(final String filename) {
        return findOrCreate(null, filename);
    }

    public static VirtualFile findOrCreate(final String directory, final String filename) {
        return ApplicationManager.getApplication().runWriteAction((Computable<VirtualFile>) () -> {
            try {
                return directory(directory).findOrCreateChildData(null, filename);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }

    static void open(final Project project, final VirtualFile scratchFile) {
        ApplicationManager.getApplication().invokeAndWait(() -> FileEditorManager.getInstance(project).openFile(scratchFile, true));
    }

    static VirtualFile directory(final String directory) {
        final ScratchFileService service = ScratchFileService.getInstance();
        final ScratchRootType rootType = ScratchRootType.getInstance();
        final String scratch = service.getRootPath(rootType);
        final Path configPath = Paths.get(String.join(File.separator, scratch, ROOT_DIRECTORY, directory));

        final VirtualFile scratchDirectory = LocalFileSystem.getInstance().findFileByNioFile(configPath);
        return ApplicationManager.getApplication().runWriteAction(
                (Computable<VirtualFile>) () -> Optional.ofNullable(scratchDirectory)
                        .orElse(mkdirs(configPath))
        );
    }

    @SneakyThrows
    static VirtualFile mkdirs(final Path path) {
        final LocalFileSystem localFileSystem = LocalFileSystem.getInstance();
        final VirtualFile file = localFileSystem.findFileByNioFile(path);
        if (Objects.nonNull(file)) {
            return file;
        }
        final Path parent = path.getParent();
        final VirtualFile parentDirectory = mkdirs(parent);
        return parentDirectory.createChildDirectory(null, path.getFileName().toString());
    }

    private static <T> @NotNull ScratchConfig getScratchConfig(final TypeReference<T> type) {
        ScratchConfig scratchConfig = null;
        if (type.getType() instanceof Class<?> tClass) {
            scratchConfig = tClass.getAnnotation(ScratchConfig.class);
        }
        if (type.getType() instanceof ParameterizedType parameterizedType) {
            scratchConfig = ((Class<?>) parameterizedType.getActualTypeArguments()[0]).getAnnotation(ScratchConfig.class);
        }
        if (Objects.isNull(scratchConfig)) {
            throw new IllegalArgumentException("ScratchConfig annotation is missing");
        }
        return scratchConfig;
    }

    private static @NotNull ScratchConfig getScratchConfig(final Type type) {
        ScratchConfig scratchConfig = null;
        if (type instanceof Class<?> tClass) {
            scratchConfig = tClass.getAnnotation(ScratchConfig.class);
        }
        if (type instanceof ParameterizedType parameterizedType) {
            scratchConfig = ((Class<?>) parameterizedType.getActualTypeArguments()[0]).getAnnotation(ScratchConfig.class);
        }
        if (Objects.isNull(scratchConfig)) {
            throw new IllegalArgumentException("ScratchConfig annotation is missing");
        }
        return scratchConfig;
    }

}

File: src/main/kotlin/com/github/morningzeng/toolbox/utils/ScratchFileUtils.kt
package com.github.morningzeng.toolbox.utils

import com.fasterxml.jackson.core.type.TypeReference
import com.github.morningzeng.toolbox.annotations.ScratchConfig
import com.github.morningzeng.toolbox.enums.JacksonType
import com.intellij.ide.scratch.ScratchFileService
import com.intellij.ide.scratch.ScratchRootType
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.newvfs.RefreshQueue
import com.intellij.util.containers.stream
import java.io.File
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import java.nio.file.Path
import java.nio.file.Paths
import java.util.function.Consumer
import java.util.stream.Stream

/**
 * @author Morning Zeng
 * @since 2025-05-16
 */
@Suppress("unused")
object ScratchFileUtils {
    private const val ROOT_DIRECTORY: String = "Toolset"

    fun <T> writeAndOpen(project: Project, filename: String, type: JacksonType, t: T) {
        write(filename, type.prettySerialize(t) ?: "") { open(project, it) }
    }

    fun <T> writeAndOpen(project: Project, filename: String, content: String) {
        write(filename, content) { open(project, it) }
    }

    fun <T> write(t: T, typeReference: TypeReference<T>) {
        t?.let {
            getConfig(typeReference).also {
                val filename = it.outputType.fullName(it.value)
                val file = findOrCreate(it.directory, filename)
                write(file, it.outputType, t)
            }
        }
    }

    fun <T> write(directory: String?, filename: String, type: JacksonType, data: T) {
        write(findOrCreate(directory, type.fullName(filename)), type, data)
    }

    fun <T> write(file: VirtualFile, type: JacksonType, data: T) = write(
        file, type.serialize(data) ?: ""
    )

    fun write(filename: String, content: String) = write(
        findOrCreate(filename), content
    )

    fun write(filename: String, content: String, consumer: Consumer<VirtualFile>) = write(
        findOrCreate(filename), content, consumer
    )

    fun write(file: VirtualFile, content: String) = write(file, content) { }

    fun write(file: VirtualFile, content: String, consumer: Consumer<VirtualFile>) {
        ApplicationManager.getApplication().invokeLater {
            ApplicationManager.getApplication().runWriteAction {
                VfsUtil.saveText(file, content)
            }
            RefreshQueue.getInstance().createSession(true, true, null)
                .also { it.addFile(file) }
                .also { it.launch() }
            consumer.accept(file)
        }
    }

    fun <T> read(clazz: Class<T>): T? = getConfig(clazz).let {
        read(it.directory, it.value, it.outputType, clazz)
    }

    fun <T> read(typeReference: TypeReference<T>): T? = getConfig(typeReference).let {
        read(it.directory, it.value, it.outputType, typeReference)
    }

    fun <T> read(typeParameter: Type): T? = getConfig(typeParameter).let {
        read(it.directory, it.value, it.outputType, typeParameter)
    }

    fun <T> read(directory: String?, filename: String, type: JacksonType, clazz: Class<T>): T? {
        return read(directory, type.fullName(filename)).let {
            type.deserialize(it, clazz)
        }
    }

    fun <T> read(directory: String?, filename: String, type: JacksonType, typeReference: TypeReference<T>): T? {
        return read(directory, type.fullName(filename)).let {
            type.deserialize(it, typeReference)
        }
    }

    fun <T> read(directory: String?, filename: String, type: JacksonType, typeParameter: Type): T? {
        return read(directory, type.fullName(filename)).let {
            type.deserialize(it, typeParameter)
        }
    }

    fun <T> read(file: VirtualFile, type: JacksonType, clazz: Class<T>): T? {
        return read(file).let {
            type.deserialize(it, clazz)
        }
    }

    fun <T> read(file: VirtualFile, type: JacksonType, typeReference: TypeReference<T>): T? {
        return read(file).let {
            type.deserialize(it, typeReference)
        }
    }

    fun <T> read(file: VirtualFile, type: JacksonType, typeParameter: Type): T? {
        return read(file).let {
            type.deserialize(it, typeParameter)
        }
    }

    fun read(directory: String?, filename: String): String {
        return read(findOrCreate(directory, filename))
    }

    fun read(file: VirtualFile): String {
        return ApplicationManager.getApplication().runReadAction<String> {
            VfsUtil.loadText(file)
        }
    }

    fun open(project: Project, directory: String?, filename: String) {
        findOrCreate(directory, filename).also {
            open(project, it)
        }
    }

    fun childrenFile(directory: String?, consumer: Consumer<Stream<VirtualFile>>) {
        ApplicationManager.getApplication().invokeLater {
            consumer.accept(directory(directory).children.stream())
        }
    }

    fun findOrCreate(filename: String): VirtualFile = findOrCreate(null, filename)

    fun findOrCreate(directory: String?, filename: String): VirtualFile {
        return ApplicationManager.getApplication().runWriteAction<VirtualFile> {
            directory(directory).findOrCreateChildData(null, filename)
        }
    }

    fun open(project: Project, file: VirtualFile) {
        ApplicationManager.getApplication().invokeAndWait {
            FileEditorManager.getInstance(project).openFile(file, true)
        }
    }

    fun directory(directory: String?): VirtualFile {
        val service = ScratchFileService.getInstance()
        val rootType = ScratchRootType.getInstance()
        val configPath = service.getRootPath(rootType).let {
            if (directory == null) {
                Paths.get("$it${File.separator}$ROOT_DIRECTORY")
            } else {
                Paths.get("$it${File.separator}$ROOT_DIRECTORY${File.separator}$directory")
            }
        }
        return ApplicationManager.getApplication().runWriteAction<VirtualFile> {
            mkdirs(configPath)
        }
    }

    fun mkdirs(path: Path): VirtualFile {
        val instance = LocalFileSystem.getInstance()
        val virtualFile = instance.findFileByNioFile(path)
        if (virtualFile != null) {
            return virtualFile
        }
        val parent = mkdirs(path.parent)
        return parent.createChildDirectory(null, path.fileName.toString())
    }

    private fun <T> getConfig(typeReference: TypeReference<T>): ScratchConfig {
        val type = typeReference.type
        return getConfig(type)
    }

    private fun getConfig(type: Type): ScratchConfig {
        if (type is Class<*>) {
            return type.getAnnotation(ScratchConfig::class.java)
        }
        if (type is ParameterizedType) {
            return (type.actualTypeArguments[0] as Class<*>).getAnnotation(ScratchConfig::class.java)
        }
        throw IllegalArgumentException("ScratchConfig annotation is missing")
    }
}